segmentstore/contracts/src/main/java/io/pravega/segmentstore/contracts/tables/TableStore.java

```diff
@@ -0,0 +1,256 @@
+     *                          {@link IteratorItem#getContinuationToken()}.
+     * @param timeout           Timeout for the operation.
+     * @return A CompletableFuture that, when completed, will return an {@link AsyncIterator} that can be used to iterate
+     * over all the {@link TableKey} instances in the given Table Segment. If the operation failed, the Future will be
+     * failed with the causing exception. Notable exceptions:
+     * <ul>
+     * <li>{@link StreamSegmentNotExistsException} If the Table Segment does not exist.
+     * <li>{@link BadSegmentTypeException} If segmentName refers to a non-Table Segment.
+     * </ul>
+     */
+    CompletableFuture<AsyncIterator<IteratorItem<TableKey>>> keyIterator(String segmentName, IteratorState continuationToken,
+                                                                         Duration timeout);
+
+    /**
+     * Creates a new Iterator over all the {@link TableEntry} instances in the given Table Segment.
+     *
+     * @param segmentName       The name of the Table Segment to iterate over.
+     * @param continuationToken An {@link IteratorState} representing a continuation token that can be used to resume a
+     *                          previously interrupted iteration. This can be obtained by invoking
+     *                          {@link IteratorItem#getContinuationToken()}.
+     * @param timeout           Timeout for the operation.
+     * @return A CompletableFuture that, when completed, will return an {@link AsyncIterator} that can be used to iterate
+     * over all the {@link TableEntry} instances in the given Table Segment. If the operation failed, the Future will be
+     * failed with the causing exception. Notable exceptions:
+     * <ul>
+     * <li>{@link StreamSegmentNotExistsException} If the Table Segment does not exist.
+     * <li>{@link BadSegmentTypeException} If segmentName refers to a non-Table Segment.
+     * </ul>
+     */
+    CompletableFuture<AsyncIterator<IteratorItem<TableEntry>>> entryIterator(String segmentName, IteratorState continuationToken,
+                                                                             Duration timeout);
+
+    /**
+     * Gets the {@link UpdateListener} for a particular TableSegment.
+     *
+     * @param listener    The {@link UpdateListener} to register.
+     * @param timeout     Timeout for the operation.
+     * @return A CompletableFuture that, when completed, will contain a {@link UpdateListener}. If the operation failed,
+     * the future will be failed with the causing exception. Notable exceptions:
+     * <ul>
+     * <li>{@link StreamSegmentNotExistsException} If the Table Segment does not exist.
+     * <li>{@link BadSegmentTypeException} If segmentName refers to a non-Table Segment.
+     * </ul>
+     */
+    CompletableFuture<Void> registerListener(UpdateListener listener, Duration timeout);
```

Comment:
An application interested in updates registers a callback using the `KeyUpdateListener` interface. The invocation is supposed to propagate via a command and register a listener against this interface. In the case the segment container changes owners, does the client lose the subscription? Is it notified?

Reply:
The client will need to re-register the update listener on the new host. Server-side all these are connection-dependent and we can't establish a connection from the server to the client.